htlcswitch/iterator.go

```diff
@@ -85,19 +89,39 @@ type HopIterator interface {
 	// EncodeNextHop encodes the onion packet destined for the next hop
 	// into the passed io.Writer.
 	EncodeNextHop(w io.Writer) error
+
+	// OnionPacket returns the original onion packet used to generate the
+	// secret keys for this instance.
+	OnionPacket() *sphinx.OnionPacket
 }
 
 // sphinxHopIterator is the Sphinx implementation of hop iterator which uses
 // onion routing to encode the payment route  in such a way so that node might
 // see only the next hop in the route..
 type sphinxHopIterator struct {
-	// nextPacket is the decoded onion packet for the _next_ hop.
-	nextPacket *sphinx.OnionPacket
+	// ogPacket is the original packet from which the processed packet is
+	// derived.
+	ogPacket *sphinx.OnionPacket
 
 	// processedPacket is the outcome of processing an onion packet. It
 	// includes the information required to properly forward the packet to
 	// the next hop.
 	processedPacket *sphinx.ProcessedPacket
+
+	// nextPacket is the decoded onion packet for the _next_ hop.
+	nextPacket *sphinx.OnionPacket
```

Comment:
Why's this field necessary? It can already be accessed via `processedPacket.nextPacket`.

Reply:
Fixed!